File: frontend/lib/liveWire.ts
// Decodes `/ws/live` snapshots (see backend/app/services/live_wire.py). The
// server sends every train as a row of values under one list of field
// names, with station names listed once. A full snapshot carries every
// field; the ticks between carry only what changes during a run
// (positions, delays, ETAs), and arrive deflated in binary frames.

import type { SnapshotCell, SnapshotTableMessage, StationRef, TrainPositionUpdate } from "./types";

type TrainField = keyof TrainPositionUpdate;

const TRAIN_FIELDS = [
  "train_id",
  "train_type",
  "service_code",
  "ac",
  "line_code",
  "line_name",
  "route_code",
  "coach_count",
  "origin",
  "destination",
  "current_station",
  "next_station",
  "direction_forward",
  "direction_label",
  "chainage_m",
  "speed_kmh",
  "delay_seconds",
  "eta_seconds",
  "destination_eta_seconds",
  "status",
  "last_updated_epoch",
] as const satisfies readonly TrainField[];

// Compile-time check that every TrainPositionUpdate field is decoded.
type UndecodedField = Exclude<TrainField, (typeof TRAIN_FIELDS)[number]>;
const everyFieldDecoded: [UndecodedField] extends [never] ? true : never = true;
void everyFieldDecoded;

const KNOWN_FIELDS: ReadonlySet<string> = new Set(TRAIN_FIELDS);
const STATION_FIELDS: ReadonlySet<TrainField> = new Set(["origin", "destination", "current_station", "next_station"]);

/** The fields a moving-parts tick carries for one train. */
export type TrainMovement = Partial<TrainPositionUpdate> & Pick<TrainPositionUpdate, "train_id">;

export type LiveSnapshot =
  | { kind: "full"; trains: TrainPositionUpdate[] }
  | { kind: "moving"; trains: TrainMovement[] };

export class SnapshotDecodeError extends Error {
  constructor(message: string) {
    super(`Malformed live snapshot: ${message}`);
    this.name = "SnapshotDecodeError";
  }
}

function isSnapshotTable(value: unknown): value is SnapshotTableMessage {
  if (typeof value !== "object" || value === null) return false;
  const message = value as Partial<SnapshotTableMessage>;
  return (
    message.type === "snapshot.table" &&
    Array.isArray(message.fields) &&
    Array.isArray(message.trains) &&
    typeof message.stations === "object" &&
    message.stations !== null
  );
}

/** The text of a frame: binary frames are raw-deflated JSON. */
export async function inflateMessage(data: string | ArrayBuffer): Promise<string> {
  if (typeof data === "string") return data;
  const stream = new Blob([data]).stream().pipeThrough(new DecompressionStream("deflate-raw"));
  return new Response(stream).text();
}

/** Whether this browser can inflate the server's compressed frames. */
export function canInflate(): boolean {
  return typeof DecompressionStream !== "undefined";
}

/**
 * The trains in a `/ws/live` message, or `null` for a message that isn't a
 * snapshot. Throws SnapshotDecodeError when a snapshot doesn't match the
 * fields this client knows.
 */
export function decodeSnapshot(message: unknown): LiveSnapshot | null {
  if (!isSnapshotTable(message)) return null;
  const full = message.full !== false;

  const columns: (readonly [TrainField, number])[] = full
    ? TRAIN_FIELDS.map((field) => {
        const index = message.fields.indexOf(field);
        if (index < 0) throw new SnapshotDecodeError(`no "${field}" column`);
        return [field, index] as const;
      })
    : message.fields.flatMap((field, index) =>
        KNOWN_FIELDS.has(field) ? [[field as TrainField, index] as const] : [],
      );
  if (!full && !columns.some(([field]) => field === "train_id")) {
    throw new SnapshotDecodeError('no "train_id" column');
  }

  const station = (cell: SnapshotCell): StationRef | null => {
    if (cell === null) return null;
    if (typeof cell !== "string") throw new SnapshotDecodeError(`station ${String(cell)} isn't a code`);
    const name = message.stations[cell];
    if (name === undefined) throw new SnapshotDecodeError(`station ${cell} isn't listed`);
    return { code: cell, name };
  };

  const rows = message.trains.map((row) => {
    if (!Array.isArray(row) || row.length !== message.fields.length) {
      throw new SnapshotDecodeError("a row doesn't match the columns");
    }
    const train: Record<string, SnapshotCell | StationRef | null> = {};
    for (const [field, index] of columns) {
      const cell = row[index] ?? null;
      train[field] = STATION_FIELDS.has(field) ? station(cell) : cell;
    }
    return train;
  });

  return full
    ? { kind: "full", trains: rows as unknown as TrainPositionUpdate[] }
    : { kind: "moving", trains: rows as unknown as TrainMovement[] };
}

File: backend/app/services/live_wire.py
"""Encodes the live snapshot for the WebSocket (see SnapshotTableMessage).

Every client receives the whole network every second, so the encoding
decides how much data a phone, and the host, spends watching trains:

- It's a table: field names once, station names once, positions rounded
  to what the map can show.
- A run's details that don't change (route, destination, coaches...) go
  out in a full snapshot on connect and every FULL_SNAPSHOT_EVERY_TICKS
  ticks; the ticks between carry only what moves.
- A train's lat/lon/heading stay off the wire: the app places trains by
  chainage along their track.
- Clients that ask for it get the table deflated, in binary frames. The
  host's edge (Cloudflare, in front of Render) strips WebSocket
  compression, so the app compresses it itself.

About 5.5 KB a second for ~200 trains, against 37 KB as a plain table.
Clients that connect without asking for this format (an app loaded before
it) get the original full table every tick.
"""

from __future__ import annotations

import json
import zlib
from collections.abc import Sequence
from typing import Literal

from app.schemas.station import StationRef
from app.schemas.train import TrainPositionUpdate
from app.schemas.websocket import Cell

WireFormat = Literal["legacy", "table", "deflate"]

FULL_SNAPSHOT_EVERY_TICKS = 10

SNAPSHOT_FIELDS: tuple[str, ...] = tuple(TrainPositionUpdate.model_fields)

STATION_FIELDS = frozenset({"origin", "destination", "current_station", "next_station"})

# Not read by the app, which places trains by chainage.
OFF_WIRE_FIELDS = frozenset({"lat", "lon", "heading_deg"})

# Fixed for a run: sent only in full snapshots.
RUN_FIELDS = frozenset(
    {
        "train_type",
        "service_code",
        "ac",
        "line_code",
        "line_name",
        "route_code",
        "coach_count",
        "origin",
        "destination",
        "direction_forward",
        "direction_label",
    }
)

FULL_FIELDS: tuple[str, ...] = tuple(name for name in SNAPSHOT_FIELDS if name not in OFF_WIRE_FIELDS)
MOVING_FIELDS: tuple[str, ...] = tuple(name for name in FULL_FIELDS if name not in RUN_FIELDS)

# Decimal places kept on the wire: ~0.1 m for coordinates and chainage,
# which is finer than anything the map draws.
ROUNDING: dict[str, int] = {
    "lat": 6,
    "lon": 6,
    "heading_deg": 1,
    "chainage_m": 1,
    "speed_kmh": 1,
    "last_updated_epoch": 2,
}


def encode_table(
    server_time_epoch: float, trains: Sequence[TrainPositionUpdate], fields: Sequence[str], *, full: bool
) -> str:
    """A SnapshotTableMessage holding `fields` of every train, as JSON text."""
    stations: dict[str, str] = {}
    rows: list[list[Cell]] = []
    for train in trains:
        row: list[Cell] = []
        for name in fields:
            value = getattr(train, name)
            if name in STATION_FIELDS:
                row.append(_station_code(value, stations))
            elif name in ROUNDING:
                row.append(round(value, ROUNDING[name]))
            else:
                row.append(value)
        rows.append(row)
    message = {
        "type": "snapshot.table",
        "server_time_epoch": round(server_time_epoch, 3),
        "full": full,
        "fields": list(fields),
        "stations": stations,
        "trains": rows,
    }
    return json.dumps(message, ensure_ascii=False, allow_nan=False, separators=(",", ":"))


def encode_snapshot(server_time_epoch: float, trains: Sequence[TrainPositionUpdate]) -> str:
    """The original format: every field of every train."""
    return encode_table(server_time_epoch, trains, SNAPSHOT_FIELDS, full=True)


def deflate(text: str) -> bytes:
    """Raw deflate, which browsers inflate with DecompressionStream("deflate-raw")."""
    compressor = zlib.compressobj(level=6, wbits=-15)
    return compressor.compress(text.encode()) + compressor.flush()


class Snapshot:
    """One tick's trains, encoded for each wire format on first use and
    then shared by every client that wants that format."""

    def __init__(self, server_time_epoch: float, trains: Sequence[TrainPositionUpdate], *, full: bool) -> None:
        self._server_time_epoch = server_time_epoch
        self._trains = trains
        self._full = full
        self._legacy: str | None = None
        self._table: str | None = None
        self._deflated: bytes | None = None

    def payload(self, wire: WireFormat) -> str | bytes:
        if wire == "legacy":
            if self._legacy is None:
                self._legacy = encode_snapshot(self._server_time_epoch, self._trains)
            return self._legacy
        if wire == "deflate":
            if self._deflated is None:
                self._deflated = deflate(self._table_text())
            return self._deflated
        return self._table_text()

    def _table_text(self) -> str:
        if self._table is None:
            fields = FULL_FIELDS if self._full else MOVING_FIELDS
            self._table = encode_table(self._server_time_epoch, self._trains, fields, full=self._full)
        return self._table


def _station_code(station: StationRef | None, names: dict[str, str]) -> str | None:
    if station is None:
        return None
    names[station.code] = station.name
    return station.code
